### PR TITLE
Upgrade GraphiQL to 1.5.1

### LINKED ISSUE #544: Subscriptions didnt work out of the box in current examples and readme.md
Maybe I miss something, but I copied the example server over to play with it and subscriptions always raise a not supported exception.

It took me some time until I noticed that this line is missing both in the example server and in the readme.md.

`services.AddSingleton<IDocumentExecuter,SubscriptionDocumentExecuter>();`

### LINKED ISSUE #588: Wrong document executer registered for subscription support
Due to the following code, `DocumentExecuter` is registered instead of `SubscriptionDocumentExecuter`, and there is nothing to override this within `.AddWebSockets()`.

https://github.com/graphql-dotnet/server/blob/4bb8cc3e87a92c179f2af8b2422fe303d51ef0c5/src/Core/Extensions/ServiceCollectionExtensions.cs#L53

This causes the following error to be thrown when attempting to use subscriptions, unless a specific implementation of `IDocumentExecuter` has been registered within the DI framework:

```
DocumentExecuter does not support executing subscriptions. You can use SubscriptionDocumentExecuter from GraphQL.Subscriptions package to execute subscriptions.
```

Current workaround is to register the correct document executer manually:

```cs
services.AddSingleton<IDocumentExecuter, SubscriptionDocumentExecuter>();
```

This is either a bug, or not properly documented and demonstrated in the readme.

See https://github.com/graphql-dotnet/graphql-dotnet/pull/1973

### PR DESCRIPTION
**Pull request by boulc**
# Context

The goal is to add support request headers in GraphiQL (discussed in https://github.com/graphql-dotnet/server/issues/504).

# Description

Request headers support has been added in recent versions of GraphiQL. Unfortunately, [GraphiQL with extensions](https://github.com/OneGraph/graphiql-with-extensions) does not support it. As it wraps the GraphiQL component and masks the props, upgrading has no effect. There is not much activity on graphiql-with-extensions repo since some time now.

The current integration of GraphiQL has a broken support of subscription (see #544), due to the deprecation of `subscribe` function in https://github.com/apollographql/subscriptions-transport-ws/pull/272. Also this dependency is still required though not [maintained anymore](https://the-guild.dev/blog/graphql-over-websockets) until #492 is resolved. For the same reason, the current integration relies on [graphiql-subscriptions-fetcher](https://github.com/apollographql/GraphiQL-Subscriptions-Fetcher) which is archived and not maintained anymore, also wrapping the `GraphQLFetcher` and masking the additional options necessary to support request headers.

# Proposal

1. Upgrading GraphiQL to version 1.5.1: cshtml file is inspired from [CDN example](https://github.com/graphql/graphiql/blob/main/examples/graphiql-cdn/index.html).

2. Adding an option to enable/disable GraphiQL with extensions: `ExplorerExtensionEnabled`, enabled by default for backward compatibility.

3. Adding an option to enable/disable header editor: `HeaderEditorEnabled`, enable by default (same as in GraphiQL).

4. Wrapping `subscriptionsFetcher` to pass additional option to support request headers editor.

5. Downgrade subscriptions-transport-ws to version 0.8.2 to fix #544.

# Tests

Screenshot of side by side example of a mutation on the left leading to a working subscription on the right:

Screenshot 2021-11-20 210857

On the previous screenshot, the request headers editor is visible on the right and the authorization header is available in HTTP context when debugging:

Screenshot 2021-11-20 210339

### CODE REVIEW

File: samples/Samples.Server/Startup.cs [hunk 1]
```diff
@@ -43,6 +43,9 @@ public void ConfigureServices(IServiceCollection services)
                 .AddWebSockets()
                 .AddDataLoader()
                 .AddGraphTypes(typeof(ChatSchema));
+
+            services
+                .AddSingleton<IDocumentExecuter, SubscriptionDocumentExecuter>();
```
Comment (sungam3r) [inline samples/Samples.Server/Startup.cs:48]: Not needed anymore.

Comment (boulc) [reply samples/Samples.Server/Startup.cs]: Right, following #653. Thanks.
Didn't it also resolve #588?

Comment (Shane32) [reply samples/Samples.Server/Startup.cs]: Looks like it to me.  I'll close that issue.

File: src/Ui.GraphiQL/GraphiQLOptions.cs [hunk 2]
```diff
@@ -35,5 +35,16 @@ public class GraphiQLOptions
         /// Gets or sets a delegate that is called after all transformations of the GraphiQL UI page.
         /// </summary>
         public Func<GraphiQLOptions, string, string> PostConfigure { get; set; } = (options, result) => result;
+
+        /// <summary>
+        /// Enables the header editor when <c>true</c>.
+        /// Not supported when ExplorerExtensionEnabled is <c>true</c>.
```
Comment (sungam3r) [inline src/Ui.GraphiQL/GraphiQLOptions.cs:41]: Why so ?

File: src/Ui.GraphiQL/GraphiQLOptions.cs [hunk 3]
```diff
@@ -35,5 +35,16 @@ public class GraphiQLOptions
...
+        /// </summary>
+        public bool HeaderEditorEnabled { get; set; } = true;
```
Comment (sungam3r) [inline src/Ui.GraphiQL/GraphiQLOptions.cs:43]:
```suggestion
        public bool HeadersEditorEnabled { get; set; } = true;
```
?

Comment (boulc) [reply src/Ui.GraphiQL/GraphiQLOptions.cs:43]: I would chose the original naming from graphiql over the plural: https://github.com/graphql/graphiql/blob/08250feb6ee8335c3b1ca83a912911ae92a75722/packages/graphiql/src/components/GraphiQL.tsx#L186

File: src/Ui.GraphiQL/GraphiQLOptions.cs (hunk 2 above)
Comment (boulc) [reply src/Ui.GraphiQL/GraphiQLOptions.cs:41]: When `ExplorerExtensionsEnabled` is true, GraphiQL with extensions is registered. This is the current existing behavior, which is maintained for backward compatibility.
Unfortunately, GraphiQL with extensions does not support the header editor.
So to enable it, you have to disable GraphiQL with extensions.

File: src/Ui.GraphiQL/GraphiQLOptions.cs (hunk 3 above)
Comment (sungam3r) [reply src/Ui.GraphiQL/GraphiQLOptions.cs:43]: OK. Please copy link to this "source-settings" into xml comments for GraphiQLOptions.

File: src/Ui.GraphiQL/GraphiQLOptions.cs (hunk 2 above)
Comment (sungam3r) [reply src/Ui.GraphiQL/GraphiQLOptions.cs:41]: OK

Comment (sungam3r) [inline src/Ui.GraphiQL/GraphiQLOptions.cs:41]:
```suggestion
        /// Not supported when <see cref="ExplorerExtensionEnabled"/> is <see langword="true"/>.
```

File: src/Ui.GraphiQL/GraphiQLOptions.cs [hunk 4]
```diff
@@ -35,5 +35,16 @@ public class GraphiQLOptions
         /// Gets or sets a delegate that is called after all transformations of the GraphiQL UI page.
         /// </summary>
         public Func<GraphiQLOptions, string, string> PostConfigure { get; set; } = (options, result) => result;
+
+        /// <summary>
+        /// Enables the header editor when <c>true</c>.
```
Comment (sungam3r) [inline src/Ui.GraphiQL/GraphiQLOptions.cs:40]:
```suggestion
        /// Enables the header editor when <see langword="true"/>.
```

File: src/Ui.GraphiQL/GraphiQLOptions.cs [hunk 5]
```diff
@@ -35,5 +35,16 @@ public class GraphiQLOptions
...
+        /// Not supported when ExplorerExtensionEnabled is <c>true</c>.
+        /// </summary>
+        public bool HeaderEditorEnabled { get; set; } = true;
+
+        /// <summary>
+        /// Enables the explorer extension when <c>true</c>.
```
Comment (sungam3r) [inline src/Ui.GraphiQL/GraphiQLOptions.cs:46]:
```suggestion
        /// Enables the explorer extension when <see langword="true"/>.
```

**Comment by boulc** [comment]
chore: code review suggestion

**Review by Shane32 — APPROVED** [review]

**Comment by sungam3r** [comment]
I need some time to check.

File: src/Ui.GraphiQL/Internal/graphiql.cshtml [hunk 6]
```diff
@@ -8,148 +8,175 @@
...
+    If you do not want to rely on a CDN, you can host these files locally or
+    include them directly in your favored resource bunder.
+  -->
+  <script
+    src="https://cdn.jsdelivr.net/npm/whatwg-fetch@2.0.3/fetch.min.js"
+    integrity="sha384-dcF7KoWRaRpjcNbVPUFgatYgAijf8DqW6NWuqLdfB5Sb4Cdbb8iHX7bHsl9YhpKa"
+    crossorigin="anonymous"
+  ></script>
+  <script
+    src="https://unpkg.com/react@16.14.0/umd/react.production.min.js"
```
Comment (sungam3r) [inline src/Ui.GraphiQL/Internal/graphiql.cshtml:38]: Why do you use SRI in some script/link tags and in some not ?

File: src/Ui.GraphiQL/Internal/graphiql.cshtml [hunk 7]
```diff
@@ -8,148 +8,175 @@
...
+  <script src="https://unpkg.com/graphiql-subscriptions-fetcher@0.0.2/browser/client.js"></script>
 
 </head>
 <body>
-	<div id="graphiql">Loading...</div>
-	<script>
+  <div id="graphiql">Loading...</div>
+  <script
+    src="https://unpkg.com/graphiql@1.5.1/graphiql.min.js"
+    type="application/javascript"
```
Comment (sungam3r) [inline src/Ui.GraphiQL/Internal/graphiql.cshtml:71]: I think that type attribute is deprecated. No?

**Comment by sungam3r** [comment]
@boulc Thanks for the fix. I don't know JS. I roughly understood what was happening here. I have two questions after which I will be glad to merge.

File: src/Ui.GraphiQL/Internal/graphiql.cshtml (hunk 6 above)
Comment (boulc) [reply src/Ui.GraphiQL/Internal/graphiql.cshtml:38]: Most of the code has been retrieved from the CDN example file of graphiql repository: https://github.com/graphql/graphiql/blob/main/examples/graphiql-cdn/index.html and actually there are no SRI in there. And for the other dependencies, I added it whenever possible. I could try to add them everywhere,in this pull request or another. What do you think?

Comment (sungam3r) [reply src/Ui.GraphiQL/Internal/graphiql.cshtml:38]: Please add them here. https://www.srihash.org/ helps a lot.

File: src/Ui.GraphiQL/Internal/graphiql.cshtml (hunk 7 above)
Comment (boulc) [reply src/Ui.GraphiQL/Internal/graphiql.cshtml:71]: Actually, the type is defaulted to `text/javascript` when not provided. This stack overflow answer seems the most accurate with references to the HTML specs per version: https://stackoverflow.com/a/4195504. I wonder why the maintainer of graphiql chose `application/javascript` over `text/javascript` 🤔.
RFC 4329 indicates that `text/javascript` mime type is obsklet though: https://datatracker.ietf.org/doc/html/rfc4329#section-7.1
I guess we could remove it, assuming all recent browsers support those specs.

Comment (boulc) [reply src/Ui.GraphiQL/Internal/graphiql.cshtml:71]: I will do some tests with various browsers and revert back.

Comment (sungam3r) [reply src/Ui.GraphiQL/Internal/graphiql.cshtml:71]: I would remove. If one really needs support for old browsers then he can use `GraphiQLOptions.PostConfigure` to inject/ovewrite any part of html.

**Comment by Shane32** [comment]
@boulc @sungam3r Status?  Waiting on this PR to release 5.2

File: src/Ui.GraphiQL/Internal/graphiql.cshtml (hunk 7 above)
Comment (boulc) [reply src/Ui.GraphiQL/Internal/graphiql.cshtml]: I removed it, without any issue 👍

File: src/Ui.GraphiQL/Internal/graphiql.cshtml (hunk 6 above)
Comment (boulc) [reply src/Ui.GraphiQL/Internal/graphiql.cshtml:38]: And done! No issue while testing it. Great tool you provided, thanks.

**Comment by boulc** [comment]
I had some issues with missing `SubscriptionDocumentExecuter` while testing my latest changes so I had to add the dependency to `GraphQL.SystemReactive` to call `AddSubscriptionDocumentExecuter` extension (see 18cfa9d).
Also, I am suggesting another tweak to the sample, changing `sentAt` attribute from Date to DateTime nullable type and defaulting it to DateTime.UtcNow so it changes each time the mutation is executed: in the example below, the subscription reflects the mutation execution more accurately without having to change the mutation input to validate:

Animation

**Comment by Shane32** [comment]
Looks good.  @sungam3r all set?

**Comment by sungam3r** [comment]
Wait a minute.

File: samples/Samples.Server/Startup.cs [hunk 8]
```diff
@@ -34,6 +35,7 @@ public void ConfigureServices(IServiceCollection services)
             services.AddSingleton<IChat, Chat>();
 
             MicrosoftDI.GraphQLBuilderExtensions.AddGraphQL(services)
+                .AddSubscriptionDocumentExecuter()
```
Comment (sungam3r) [inline samples/Samples.Server/Startup.cs:38]: Examples are "completely" fixed in #693 and will be merged for v6.

**Review by sungam3r — APPROVED** [review]

File: samples/Samples.Server/Startup.cs (hunk 8 above)
Comment (boulc) [reply samples/Samples.Server/Startup.cs:38]: Oh nice! 💯
I anticipate a slight conflict with this addition but nothing major.

Comment (sungam3r) [reply samples/Samples.Server/Startup.cs:38]: Yes. That is why I needed an extra time for review.
